pkg/nsxt/nsxt_secret.go

```diff
@@ -0,0 +1,152 @@
+		}
+	}
+	return secretManager
+}
+
+// InitializeSecretManager hooks up secret informer add, update, delete callbacks
+func InitializeSecretManager(secretInformer v1.SecretInformer) error {
+	if secretManager == nil || secretManager.secretName == "" || secretManager.secretNamespace == "" {
+		klog.V(6).Infof("No need to initialize NSXT secret manager as secret is not provided")
+		return nil
+	}
+	if secretInformer == nil {
+		return errors.New("failed to initialize NSXT secret manager as secret informer is nil")
+	}
+
+	secretInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    secretManager.secretAdded,
+		UpdateFunc: secretManager.secretUpdated,
+		DeleteFunc: secretManager.secretDeleted,
+	})
+
+	return nil
+}
+
+func (nsm *SecretManager) addNsxtConnecter(connector *client.RestConnector) {
+	nsm.connectors = append(nsm.connectors, connector)
+}
+
+func (nsm *SecretManager) isForNsxtSecret(secret *corev1.Secret) bool {
+	if secret.GetName() == nsm.secretName && secret.GetNamespace() == nsm.secretNamespace {
+		return true
+	}
+	return false
+}
+
+func (nsm *SecretManager) secretAdded(obj interface{}) {
+	secret, ok := obj.(*corev1.Secret)
+	if secret == nil || !ok {
+		klog.Warningf("secretAdded: unrecognized object %+v", obj)
+		return
+	}
+	if nsm.isForNsxtSecret(secret) {
+		nsm.updateConnectorContext(secret)
+	}
+}
+
+func (nsm *SecretManager) secretUpdated(oldObj, newObj interface{}) {
+	oldSecret, ok := oldObj.(*corev1.Secret)
+	if oldSecret == nil || !ok {
+		klog.Warningf("secretUpdated: unrecognized object %+v", oldObj)
+		return
+	}
+	newSecret, ok := newObj.(*corev1.Secret)
+	if newSecret == nil || !ok {
+		klog.Warningf("secretUpdated: unrecognized object %+v", newObj)
+		return
+	}
+	if nsm.isForNsxtSecret(newSecret) && !reflect.DeepEqual(oldSecret.Data, newSecret.Data) {
+		nsm.updateConnectorContext(newSecret)
+	}
+}
+
+func (nsm *SecretManager) secretDeleted(obj interface{}) {
+	secret, ok := obj.(*corev1.Secret)
+	if secret == nil || !ok {
+		klog.Warningf("secretDeleted: unrecognized object %+v", obj)
+		return
+	}
+	if nsm.isForNsxtSecret(secret) {
+		nsm.resetConnectorContext()
+	}
+}
+
+func (nsm *SecretManager) updateConnectorContext(secret *corev1.Secret) {
+	user, password := "", ""
+	for key, value := range secret.Data {
+		if key == "username" {
+			user = string(value)
+		}
+		if key == "password" {
+			password = string(value)
+		}
+	}
+	if user == "" || password == "" {
+		klog.Warningf("NSXT username and password should be both provided in secret")
+		return
+	}
+	klog.V(6).Infof("Updating security context for NSXT connection")
+	securityCtx := core.NewSecurityContextImpl()
+	securityCtx.SetProperty(security.AUTHENTICATION_SCHEME_ID, security.USER_PASSWORD_SCHEME_ID)
+	securityCtx.SetProperty(security.USER_KEY, user)
+	securityCtx.SetProperty(security.PASSWORD_KEY, password)
+	for _, connector := range nsm.connectors {
+		connector.SetSecurityContext(securityCtx)
```

Comment:
Do you think we need to support per-host Secrets? Or do you think it's sufficient to assume a Secret is valid for every  NSX host?

Reply:
I don't think we could assume user will set the same user/password for all NSX hosts. 
For each vsphere-cloud-provider, it only has one NSXT config with one Secret, right? Or do you mean we support multiple NSXT config?

Reply:
We support multiple vCenters if a cluster could possibly have VMs from mulitple vCenters. Is that scenario likely or possible with NSX? Seems less likely but just double checking

Reply:
To have K8s nodes in the same overlay subnet, we should attach VM vNIC to the same segment in one NSX. For VMs in different vCenters, as long as their ESX hosts register to the same NSX, it could still work. So I think every K8s cluster has only one NSX.